File: assembly-adapter/src/main/java/me/panpf/adapter/ClickListenerManager.java
package me.panpf.adapter;

import android.support.annotation.IdRes;
import android.support.annotation.NonNull;
import android.view.View;

import java.util.LinkedList;
import java.util.List;

public class ClickListenerManager<DATA> {
    @NonNull
    private List<Object> holders = new LinkedList<Object>();

    public void add(@IdRes int viewId, @NonNull OnClickListener<DATA> onClickListener) {
        holders.add(new ClickListenerHolder<DATA>(viewId, onClickListener));
    }

    public void add(@NonNull OnClickListener<DATA> onClickListener) {
        holders.add(new ClickListenerHolder<DATA>(onClickListener));
    }

    public void add(@IdRes int viewId, @NonNull OnLongClickListener<DATA> onClickListener) {
        holders.add(new LongClickListenerHolder<DATA>(viewId, onClickListener));
    }

    public void add(@NonNull OnLongClickListener<DATA> onClickListener) {
        holders.add(new LongClickListenerHolder<DATA>(onClickListener));
    }

    public void register(@NonNull final ItemFactory<DATA> itemFactory, @NonNull Item<DATA> item, @NonNull View itemView) {
        for (final Object holder : holders) {
            if (holder instanceof ClickListenerHolder) {
                final ClickListenerHolder<DATA> clickListenerHolder = (ClickListenerHolder<DATA>) holder;
                int viewId = clickListenerHolder.getViewId();
                final View targetView = viewId > 0 ? itemView.findViewById(viewId) : itemView;
                if (targetView == null) {
                    throw new IllegalArgumentException("Not found target view by id " + viewId);
                }

                targetView.setTag(R.id.aa_item_holder, item);
                targetView.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        Item<DATA> item = (Item<DATA>) targetView.getTag(R.id.aa_item_holder);
                        int position = item.getAdapterPosition();
                        AssemblyAdapter adapter = itemFactory.getAdapter();
                        int positionInPart = adapter != null ? adapter.getPositionInPart(position) : position;
                        clickListenerHolder.getListener().onViewClick(v.getContext(), v, position, positionInPart, item.getData());
                    }
                });
            } else if (holder instanceof LongClickListenerHolder) {
                final LongClickListenerHolder<DATA> longClickListenerHolder = (LongClickListenerHolder<DATA>) holder;
                int viewId = longClickListenerHolder.getViewId();
                final View targetView = viewId > 0 ? itemView.findViewById(viewId) : itemView;
                if (targetView == null) {
                    throw new IllegalArgumentException("Not found target view by id " + viewId);
                }

                targetView.setTag(R.id.aa_item_holder, item);
                targetView.setOnLongClickListener(new View.OnLongClickListener() {
                    @Override
                    public boolean onLongClick(View v) {
                        Item<DATA> item = (Item<DATA>) targetView.getTag(R.id.aa_item_holder);
                        int position = item.getAdapterPosition();
                        AssemblyAdapter adapter = itemFactory.getAdapter();
                        int positionInPart = adapter != null ? adapter.getPositionInPart(position) : position;
                        return longClickListenerHolder.getListener().onViewLongClick(v.getContext(), v, position, positionInPart, item.getData());
                    }
                });
            }
        }
    }
}

File: assemblyadapter-common-item/src/main/java/com/github/panpf/assemblyadapter/internal/ClickListenerManager.kt
package com.github.panpf.assemblyadapter.internal

import androidx.annotation.IdRes
import com.github.panpf.assemblyadapter.OnClickListener
import com.github.panpf.assemblyadapter.OnLongClickListener
import java.util.*

class ClickListenerManager<DATA> {

    val holders: MutableList<Any> = LinkedList()

    fun add(@IdRes viewId: Int, onClickListener: OnClickListener<DATA>) {
        holders.add(ClickListenerHolder(viewId, onClickListener))
    }

    fun add(onClickListener: OnClickListener<DATA>) {
        holders.add(ClickListenerHolder(onClickListener))
    }

    fun add(@IdRes viewId: Int, onClickListener: OnLongClickListener<DATA>) {
        holders.add(LongClickListenerHolder(viewId, onClickListener))
    }

    fun add(onClickListener: OnLongClickListener<DATA>) {
        holders.add(LongClickListenerHolder(onClickListener))
    }

    class ClickListenerHolder<DATA>(
        @field:IdRes @param:IdRes @get:IdRes val viewId: Int,
        val listener: OnClickListener<DATA>
    ) {
        constructor(listener: OnClickListener<DATA>) : this(0, listener)
    }

    class LongClickListenerHolder<DATA>(
        @field:IdRes @param:IdRes @get:IdRes val viewId: Int,
        val listener: OnLongClickListener<DATA>
    ) {
        constructor(listener: OnLongClickListener<DATA>) : this(0, listener)
    }
}